src/Kernel32.Shared/Kernel32+OVERLAPPED.cs

```diff
@@ -0,0 +1,74 @@
+            /// <para>The Internal member was originally reserved for system use and its behavior may change.</para>
+            /// </summary>
+            public IntPtr Internal;
+
+            /// <summary>
+            /// The number of bytes transferred for the I/O request. The system sets this member if the request is completed without errors.
+            /// <para>The InternalHigh member was originally reserved for system use and its behavior may change.</para>
+            /// </summary>
+            public IntPtr InternalHigh;
+
+            /// <summary>
+            /// The low-order portion of the file position at which to start the I/O request, as specified by the user.
+            /// <para>This member is nonzero only when performing I/O requests on a seeking device that supports the concept of an offset(also referred to as a file pointer mechanism), such as a file.Otherwise, this member must be zero.</para>
+            /// <para>For additional information, see Remarks.</para>
+            /// </summary>
+            public int Offset;
+
+            /// <summary>
+            /// The high-order portion of the file position at which to start the I/O request, as specified by the user.
+            /// <para>This member is nonzero only when performing I/O requests on a seeking device that supports the concept of an offset(also referred to as a file pointer mechanism), such as a file.Otherwise, this member must be zero.</para>
+            /// <para>For additional information, see Remarks.</para>
+            /// </summary>
+            public int OffsetHigh;
+
+            /// <summary>
+            /// A handle to the event that will be set to a signaled state by the system when the operation has completed. The user must initialize this member either to zero or a valid event handle using the CreateEvent function before passing this structure to any overlapped functions. This event can then be used to synchronize simultaneous I/O requests for a device. For additional information, see Remarks.
+            /// <para>Functions such as ReadFile and WriteFile set this handle to the nonsignaled state before they begin an I/O operation.When the operation has completed, the handle is set to the signaled state.</para>
+            /// <para>Functions such as GetOverlappedResult and the synchronization wait functions reset auto-reset events to the nonsignaled state. Therefore, you should use a manual reset event; if you use an auto-reset event, your application can stop responding if you wait for the operation to complete and then call GetOverlappedResult with the bWait parameter set to TRUE.</para>
+            /// </summary>
+            public IntPtr hEvent;
```

Comment:
I don't think so, there is no marshaling involved, actually OVERLAPPED must not be marshaled (And it's passed as an `OVERLAPPED*` for this reason).

It must be the same memory in our program memory space as visible by kernel32 because when an async IO finishes the memory originally passed to `ReadFile` or `WriteFile` will be written without any involvement of the CLR at all.

It's also one of the reason why the buffer is a `void*` instead of a `byte[]` to indicate to callers that (at least in the async IO case) this memory (buffer & `OVERLAPPED` structure) absolutely need to be pinned, because it will be written any time between the call and when GetOverlappedResult  is called so the GC is not allowed to move it during this time.

It's also why I backtracked from making `OVERLAPPED` a class, there is no way for this type to be marshaled.

**Note:** `NativeOverlapped` is fun because it's much more special. The CLR actually know about it, allocate it voluntarily too big to stuff some info at the end to allow to avoid GC pining [nativeoverlapped.h L42](https://github.com/dotnet/coreclr/blob/master/src/vm/nativeoverlapped.h#42) 

The very fun part is that WCF is actually aware of this and using direct pointer manipulation into the undocumented memory to do some stuff (The comment in the file point that out [nativeoverlapped.h L32](https://github.com/dotnet/coreclr/blob/master/src/vm/nativeoverlapped.h#32) )